libr/asm/arch/8051/8051_ass.c

```diff
@@ -545,14 +545,76 @@ static bool singlearg_reladdr(ut8 const firstbyte, char const* arg
 static bool singlearg_direct(ut8 const firstbyte, char const* arg
 	, ut8 **out)
 {
-	ut8 address;
-	if (!address_direct (arg, &address)) {
-		return false;
+	bool ret = true;
+	ut8 address = 0x00;
+	if (!address_direct (arg, &address) && !is_reg (arg)) {
+		ret = false;
+	} else if (is_reg (arg)) {
```

Comment:
Move into a separate function and use a steuct snd a for loop to iterate over all the strings . Something like address = getreg(“r0”);